### PR TITLE
pub/sub rabbitmq: add publishing retries

### PR DESCRIPTION
**Pull request by dmitsh**
# Description

add publishing retries in RabbitMQ pib/sub

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #1371

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pubsub/rabbitmq/rabbitmq.go [hunk 1]
```diff
@@ -205,18 +207,26 @@ func (r *rabbitMQ) publishSync(req *pubsub.PublishRequest) (rabbitMQChannelBroke
 func (r *rabbitMQ) Publish(req *pubsub.PublishRequest) error {
 	r.logger.Debugf("%s publishing message to %s", logMessagePrefix, req.Topic)
 
-	channel, connectionCount, err := r.publishSync(req)
-	if err != nil {
+	attempt := 0
```
Comment (artursouza) [inline pubsub/rabbitmq/rabbitmq.go:210]: I am debating if we should have a flag or not to enable/disable retries in component metadata. The upcoming resiliency work, this flag would be deprecated soon after.

Another option is not to merge this and wait for the resiliency.

Comment (dmitsh) [reply pubsub/rabbitmq/rabbitmq.go:210]: We promised this fix in 1.6. I would suggest adding `publishMaxRetries` flag to RabbitMQ metadata and setting it by default to `3`.
Once resiliency is merged, we will deprecate it.
/cc @pkedy @yaron2

Comment (yaron2) [reply pubsub/rabbitmq/rabbitmq.go:210]: This work will probably conflict with resiliency.

Things like connection establishment for the first time *could* be retried inside the component because these are unlikely to be covered by the resiliency feature, but I agree with @artursouza that IF there's overlap, I'm inclined to keep this open for now and wait.

@dmitsh we could do as you suggest, but adding a flag for several weeks just to deprecate it after one release in a beta/stable level component seems unnecessary to me.

Comment (dmitsh) [reply pubsub/rabbitmq/rabbitmq.go:210]: So, if I can summarize, there are 3 options:
 (1) do nothing now, and integrate with the resiliency feature later. In the meantime, advise users to implement retries in their application.
 (2) add hard-coded retry logic (implemented in this PR), which will be later replaced with the resiliency feature
 (3) add temporary configuration for publishing retries, which will be later deprecated once the resiliency feature is merged.
 
I agree that the third option is not desired.

I would personally tend to the second option, because it unblocks current users facing this issue, and could be seamlessly replaced with the resilience feature (regardless of when it will be delivered).

Comment (yaron2) [reply pubsub/rabbitmq/rabbitmq.go:210]: > So, if I can summarize, there are 3 options: (1) do nothing now, and integrate with the resiliency feature later. In the meantime, advise users to implement retries in their application. (2) add hard-coded retry logic (implemented in this PR), which will be later replaced with the resiliency feature (3) add temporary configuration for publishing retries, which will be later deprecated once the resiliency feature is merged.
> 
> I agree that the third option is not desired.
> 
> I would personally tend towards the second option, because it unblocks current users facing this issue, and could be seamlessly replaced with the resilience feature (regardless of when it will be delivered).

Thanks for the summary. I actually vote for (2), but without removing it when resiliency is in, and here's why:

Resiliency is an opt-in feature. If we put the retries now, users will get a working component for 1.6, then in 1.7 they will experience a regression to 1.5 behavior UNLESS they specify a resiliency policy. users will not be able to possibly know they should be doing that, because they have no idea we removed *default* retries in place of optional ones. Also, the resiliency configuration to match the deprecated component retries are nearly impossible to reason about from a user perspective.

This is a perfectly valid example of keeping component retries to deal with a *known* issue and provide component level retries to deal with that.

Then, resiliency policies naturally fit in with whatever behavior the component is exhibiting with the component level retries included.
In other words, the component has default retries to guarantee an acceptable level of performance. Then, developers may want to put resiliency features to account for extra retries because their RabbitMQ environment needs it (based on configuration, hardware, resource constraints, etc.)

cc @artursouza

Comment (dmitsh) [reply pubsub/rabbitmq/rabbitmq.go:210]: > Thanks for the summary. I actually vote for (2) as well

Do you want to approve this PR?

Comment (yaron2) [reply pubsub/rabbitmq/rabbitmq.go:210]: > > Thanks for the summary. I actually vote for (2) as well
> 
> Do you want to approve this PR?

I'd like to get @artursouza's take on this approach too.

Comment (artursouza) [reply pubsub/rabbitmq/rabbitmq.go:210]: LGTM

**Review by artursouza — APPROVED** [review]
